cheatsheets/Nodejs_security_cheat_sheet.md

```diff
@@ -0,0 +1,532 @@
+
+```JavaScript
+var expectCt = require('expect-ct');
+app.use(expectCt({ maxAge: 123 }));
+app.use(expectCt({ enforce: true, maxAge: 123 }));
+app.use(expectCt({ enforce: true, maxAge: 123, reportUri: 'http://example.com'}));
+```
+
+- **Public-Key-Pins:** This header increases the security of HTTPS. With this header, a specific cryptographic public key is associated with a specific web server. If the server does not use the pinned keys in future, the browser regards the responses as illegitimate. It has 2 optional (`reportUri`, `includeSubDomains`) and 2 required (`pin-sha256`, `max-age`) directives. These can be used as follows:
+
+```JavaScript
+app.use(helmet.hpkp({
+    maxAge: 123,
+    sha256s: ['Ab3Ef123=', 'ZyxawuV45='],
+    reportUri: 'http://example.com',
+    includeSubDomains: true
+}));
+```
+
+As discussed in [Transport Layer Security Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Transport_Layer_Protection_Cheat_Sheet.html#consider-using-public-key-pinning), the decision to use public key pinning should be made with careful consideration, since it may cause locking out users for a long time if used incorrectly.
+
+- **X-Powered-By:** X-Powered-By header is used to inform what technology is used in the server side. This is an unnecessary header causing information leakage, so it should be removed from your application. To do so, you can use the `hidePoweredBy` as follows:
+
+```JavaScript
+app.use(helmet.hidePoweredBy());
+```
+
+Also, you can lie about the technologies used with this header. For example, even if your application does not use PHP, you can set X-Powered-By header to seem so.
+
+```JavaScript
+app.use(helmet.hidePoweredBy({ setTo: 'PHP 4.2.0' }));
+```
+
+## Platform Security
+
+### Keep your packages up-to-date
+
+Security of your application depends directly on how secure the third-party packages you use in your application are. Therefore, it is important to keep your packages up-to-date. It should be noted that [Using Components with Known Vulnerabilities](https://www.owasp.org/index.php/Top_10-2017_A9-Using_Components_with_Known_Vulnerabilities) is still in the OWASP Top 10. You can use [OWASP Dependency-Check](https://jeremylong.github.io/DependencyCheck/analyzers/nodejs.html) to see if any of the packages used in the project has a knwon vulnerability. Also you can use [Retire.js](https://github.com/retirejs/retire.js/) to check JavaScript libraries with known vulnerabilities. In order to use it, you can run the following commands in the source code folder of your application:
```

Comment:
For more tools we can link to: https://cheatsheetseries.owasp.org/cheatsheets/Vulnerable_Dependency_Management_Cheat_Sheet.html#tools.